abci/app.go

```diff
@@ -0,0 +1,378 @@
+		return nil, err
+	}
+	cfg, err := LoadConfig(appConfigPath)
+	if err != nil {
+		cfg = new(Config)
+		cfg.CurseWords = "bad"
+	}
+	return &ForumApp{
+		state:              loadState(db),
+		valAddrToPubKeyMap: make(map[string]cryptoproto.PublicKey),
+		CurseWords:         cfg.CurseWords,
+	}, nil
+
+}
+
+// Return application info
+func (app ForumApp) Info(_ context.Context, info *abci.RequestInfo) (*abci.ResponseInfo, error) {
+
+	if len(app.valAddrToPubKeyMap) == 0 && app.state.Height > 0 {
+		validators := app.getValidators()
+		for _, v := range validators {
+			pubkey, err := cryptoencoding.PubKeyFromProto(v.PubKey)
+			if err != nil {
+				panic(fmt.Errorf("can't decode public key: %w", err))
+			}
+			app.valAddrToPubKeyMap[string(pubkey.Address())] = v.PubKey
+		}
+	}
+	return &abci.ResponseInfo{
+		Version:         version.ABCIVersion,
+		AppVersion:      ApplicationVersion,
+		LastBlockHeight: app.state.Height,
+
+		LastBlockAppHash: app.state.Hash(),
+	}, nil
+}
+
+// Query blockchain
+func (app ForumApp) Query(ctx context.Context, query *abci.RequestQuery) (*abci.ResponseQuery, error) {
+	resp := abci.ResponseQuery{Key: query.Data}
+	// Just testing whether validators are properly stored
+	if query.Path == "/val" {
+		for k := range app.valAddrToPubKeyMap {
+			return &types.ResponseQuery{
+				Key:   query.Data,
+				Value: []byte(string(k)),
+			}, nil
+		}
+		return &types.ResponseQuery{}, nil
+	}
+	// Parse sender from query data
+	sender := string(query.Data)
+
+	if sender == "history" {
+		messages, err := model.FetchHistory(app.state.DB)
+		if err != nil {
+			return nil, err
+		}
+		resp.Log = messages
+		resp.Value = []byte(messages)
+
+		return &resp, nil
+	}
+	// Retrieve all message sent by the sender
+	messages, err := model.GetMessagesBySender(app.state.DB, sender)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert the messages to JSON and return as query result
+	resultBytes, err := json.Marshal(messages)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.Log = string(resultBytes)
+	resp.Value = resultBytes
+
+	return &resp, nil
+}
+
+func (app ForumApp) CheckTx(ctx context.Context, checktx *abci.RequestCheckTx) (*abci.ResponseCheckTx, error) {
+	// Parse the tx message
+	msg, err := model.ParseMessage(checktx.Tx)
+	if err != nil {
+		fmt.Printf("failed to parse transaction message checktx: %v\n", err)
+		return &abci.ResponseCheckTx{Code: CodeTypeInvalidTxFormat, Log: "Invalid transaction format"}, err
+	}
+	fmt.Println("Searching for sender ... ", msg.Sender)
+	u, err := app.state.DB.FindUserByName(msg.Sender)
+
+	if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+		fmt.Println("problem in check tx: ", string(checktx.Tx))
+		return &types.ResponseCheckTx{Code: CodeTypeEncodingError}, nil
+	}
+
+	if err == badger.ErrKeyNotFound {
+		fmt.Println("Not found user :", msg.Sender)
+	}
+	if u != nil && u.Banned {
+		return &types.ResponseCheckTx{Code: CodeTypeBanned, Log: "User is banned"}, nil
+	}
```

Comment:
```suggestion
	if err != nil {
		if !errors.Is(err, badger.ErrKeyNotFound) {
			fmt.Println("problem in check tx: ", string(checktx.Tx))
			return &types.ResponseCheckTx{Code: CodeTypeEncodingError}, nil
		}
		fmt.Println("Not found user :", msg.Sender)
	} else {
		if u != nil && u.Banned {
			return &types.ResponseCheckTx{Code: CodeTypeBanned, Log: "User is banned"}, nil
		}
	}
```
I find these conditions hard to follow, and hard to convince me they are exhaustive. This is a suggestion to make conditions clearer